### PR TITLE
Test/e2e add friends

### PR DESCRIPTION
**Pull request by Lil-Noe**
This pull request includes various changes to test files and components, focusing on improving test identifiers, refactoring imports, and enhancing the functionality of components. The most important change includes writing a new end-to-end test file, that follows a userflow including friends.

### Test Identifier Updates:
* Updated test IDs in `requestList.test.tsx` to use meaningful names instead of indices. [[1]](diffhunk://#diff-16a7737cb1ea4ef4cdec8e4cc976ff5f841122754eabf1ca062105c70ac7b2c6L47-R48) [[2]](diffhunk://#diff-16a7737cb1ea4ef4cdec8e4cc976ff5f841122754eabf1ca062105c70ac7b2c6L68-R69)
* Changed test IDs in `listOfFilteredUsers.test.tsx` and `userListItem.test.tsx` to use `handle-button` instead of `add-button`. [[1]](diffhunk://#diff-105b7a79e7b544c17702d4badcf7dca402310ff8f2008d13265619b9b16adee2L73-R73) [[2]](diffhunk://#diff-4246d21a0152b22c29e2e4c9cd20acaa78cb128edf7b3a1643c24407730e3629L52-R52)
* Modified test IDs in `groupIcon.test.tsx` to include the group name.

### Component Import Refactoring:
* Refactored default imports of `Timer` to named imports in various files. [[1]](diffhunk://#diff-586face8e6fff88184a7e645056c9c1d2c51dde84a4e6d80022ce08fb5f43f3cL3-R3) [[2]](diffhunk://#diff-81538be14e936d916bbb75701d18142c9f3231c8f938a58aa982bb04582cb48eL3-R3) [[3]](diffhunk://#diff-03db51d53611f74dc3d6aa2c6ebf828b61f9c6afe65f708a594d0e2e51a0f3d7L4-R4)
* Changed the import of `SearchBar` to a named function and added a test ID to its input field.

### Component Enhancements:
* Added `user` prop to `FriendsScreen` and updated related tests to pass this prop. [[1]](diffhunk://#diff-69ed67d96090317da0d8240062a8a947ea8c1f92358cd52bd100a17da185d12fR42-R48) [[2]](diffhunk://#diff-69ed67d96090317da0d8240062a8a947ea8c1f92358cd52bd100a17da185d12fL51-L57) [[3]](diffhunk://#diff-69ed67d96090317da0d8240062a8a947ea8c1f92358cd52bd100a17da185d12fR75) [[4]](diffhunk://#diff-69ed67d96090317da0d8240062a8a947ea8c1f92358cd52bd100a17da185d12fR101) [[5]](diffhunk://#diff-69ed67d96090317da0d8240062a8a947ea8c1f92358cd52bd100a17da185d12fR126) [[6]](diffhunk://#diff-69ed67d96090317da0d8240062a8a947ea8c1f92358cd52bd100a17da185d12fR167) [[7]](diffhunk://#diff-69ed67d96090317da0d8240062a8a947ea8c1f92358cd52bd100a17da185d12fR193)
* Added a test ID to the `Timer` component in `ChallengeDescription`.
* Updated `SingleComment` component to include a test ID for the comment container.

### Code Cleanup:
* Removed console logs from `ListOfFilteredUsersViewModel` and `HomeScreenViewModel`. [[1]](diffhunk://#diff-6a9f2de792f40e47b46b8fa41ea53146e08005b78b65d38d7c6a24c076d9484bL48) [[2]](diffhunk://#diff-6a9f2de792f40e47b46b8fa41ea53146e08005b78b65d38d7c6a24c076d9484bL62) [[3]](diffhunk://#diff-e004c42f6f76672db0c8b2a09814662a5171921445c50ce213c530615e256c69L85)
* Deleted the entire `groupScreen.unit.test.tsx` file, which contained outdated and redundant test cases.

### CODE REVIEW

File: app/src/views/components/friends/search_bar.tsx
Comment (Roman-batut) [inline app/src/views/components/friends/search_bar.tsx:1]: I'm unsure why we would need SearchBar to be a function as all it does as a function how you described it is returning a const. Ultimately isn't it a const ?

File: app/src/views/components/friends/request_list.tsx [hunk 1]
```diff
@@ -34,7 +34,7 @@ export function RequestList({
         <FriendRequestItem
           name={item.name}
           key={item.uid}
-          testID={index.toString()}
+          testID={item.name}
```
Comment (Roman-batut) [inline app/src/views/components/friends/request_list.tsx:37]: I think this is a better way of doing it as indexes can get messy in case of problems, so thank you !

File: app/src/views/components/challenge/timer.tsx
Comment (Roman-batut) [inline app/src/views/components/challenge/timer.tsx:1]: Could you enlighten me on what this change produces of a difference ?

File: app/src/viewmodels/components/friends/ListOfFilteredUsersViewModel.tsx
Comment (Roman-batut) [inline app/src/viewmodels/components/friends/ListOfFilteredUsersViewModel.tsx:1]: Thank you for this, it's such a pain debugging a specific feature with the console being flooded by unnecessary stuff.

File: __tests__/View-tests/friends/friendsScreen.test.tsx
Comment (Roman-batut) [inline __tests__/View-tests/friends/friendsScreen.test.tsx:1]: Great implementation of mockUser, also getAuth is an unused import you might want to delete it !

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 2]
```diff
@@ -0,0 +1,418 @@
+import React from "react";
+import {
+  render,
+  fireEvent,
+  waitFor,
+  cleanup,
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:6]: This is unused, remove it.

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+// Mock FirestoreCtrl
+jest.mock("@/src/models/firebase/FirestoreCtrl", () => {
+  return jest.fn().mockImplementation(() => {
+    return {
+      // Mock functions used in home screen
+      getGroupsByUserId: jest.fn((id) => {
+        if (id === "123") {
+          return new Promise<DBGroup[]>((resolve) => {
+            resolve([mockTesterGroup]);
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:32]: You can actually use `Promise.resolve()`instead of this more complex version.

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+          });
+        } else if (id === "456") {
+          return new Promise<DBGroup[]>((resolve) => {
+            resolve([mockTesterFriendGroup]);
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:36]: Here as well.

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+          });
+        }
+      }),
+      getChallengeDescription: jest.fn((id) => {
+        return mockCurrentChallenge;
+      }),
+      getPostsByChallengeTitle: jest.fn((title) => {
+        return new Promise<DBChallenge[]>((resolve) => {
+          resolve([mockPostTesterFriend]);
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:45]: Here as well.

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+          if (uid === "456") {
+            return true;
+          }
+        }
+      }),
+
+      // Mock functions used in maximize screen
+      getCommentsOf: jest.fn((challenge_id) => {
+        return new Promise<DBComment[]>((resolve) => {
+          resolve(mockPostComments);
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:102]: Here as well.

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+        });
+      }),
+      addComment: jest.fn(() => {}),
+      updateLikesOf: jest.fn((challenge_id, likes) => {
+        mockPostLikes = likes;
+      }),
+      getLikesOf: jest.fn((challenge_id) => {
+        return new Promise<string[]>((resolve) => {
+          resolve(mockPostLikes);
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:111]: Here as well.

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+      }),
+
+      getUser: jest.fn((uid) => {
+        if (uid === "123") {
+          return new Promise<DBUser[]>((resolve) => {
+            resolve([mockTester]);
+          });
+        } else if (uid === "456") {
+          return new Promise<DBUser[]>((resolve) => {
+            resolve([mockTesterFriend]);
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:122]: Here as well.

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+          }
+        }
+      }),
+      isRequested: jest.fn((uid, friendUid) => {
+        if (!isRequestSent) {
+          return false;
+        } else {
+          if (uid === "456") {
+            return true;
+          }
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:95]: The mock isn't fully complete as this method has some edge-cases that return nothing : `isRequestSent=true` and `uid != 456`

File: __tests__/e2e/AddFriendsStateFlow.test.tsx [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+        if (uid === "456") {
+          return [mockTester];
+        } else return [];
+      }),
+      isFriend: jest.fn((uid, friendUid) => {
+        if (!isRequestSent) {
+          return false;
+        } else {
+          if (mockTester.friends.includes(friendUid)) {
+            return true;
```
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:85]: Same comment as `isRequested`

File: __tests__/e2e/AddFriendsStateFlow.test.tsx
Comment (Roman-batut) [inline __tests__/e2e/AddFriendsStateFlow.test.tsx:1]: This file appears to be well-structured, covering essential user interactions such as sending and accepting friend requests, and commenting on posts.
The use of mock implementations for Firestore controller methods is appropriate for isolating test cases.
Well done !

File: app/src/views/components/friends/search_bar.tsx
Comment (Lil-Noe) [reply app/src/views/components/friends/search_bar.tsx:1]: It sticks to the convention of other components

File: app/src/views/components/challenge/timer.tsx
Comment (Lil-Noe) [reply app/src/views/components/challenge/timer.tsx:1]: This sticks to the convention of other components regarding default functions

File: __tests__/e2e/AddFriendsStateFlow.test.tsx (hunk 3 above)
Comment (Lil-Noe) [reply __tests__/e2e/AddFriendsStateFlow.test.tsx:32]: This way is more complex but it forces the type, which helps mocking the functions

File: __tests__/e2e/AddFriendsStateFlow.test.tsx (hunk 9 above)
Comment (Lil-Noe) [reply __tests__/e2e/AddFriendsStateFlow.test.tsx:95]: Well mocking the full function is not necessary here, and otherwise the mocked function does nothing.

**Review by Tristan-C789 — APPROVED** [review]
Thank you for such an essential feature at such a critical time! LGTM!
